File: apps/web/src/lib/share.ts
import type { UserAvatar } from "@/hooks/use-curriculum";

// Share ranking/streak ke sosmed. Link share = /s/{nickname}/{kind}: halaman
// undangan publik (tanpa login) + gambar kartu di /s/{nickname}/{kind}/image
// (dipakai juga sebagai og:image). Datanya diambil dari server lewat
// nickname, bukan dari URL, jadi rank gak bisa dipalsuin.
// Kontrak: services/api internal/curriculumsvc/share.go.

export type ShareKind = "sd" | "smp" | "smk" | "kampus" | "adventure" | "streak";

export const SHARE_KINDS: ShareKind[] = ["sd", "smp", "smk", "kampus", "adventure", "streak"];

type ShareRank = { rank: number; total: number };

export type ShareProfile = {
  nickname: string;
  avatar: UserAvatar;
  currentStreak: number;
  longestStreak: number;
  tiers: { code: string; name: string; points: number; rank: ShareRank | null }[];
  adventure: {
    checkpointsCleared: number;
    totalCheckpoints: number;
    completed: boolean;
    timePercent: number;
    rank: ShareRank;
  } | null;
};

// Isi kartu: `big` = angka utama, `caption` = keterangan di bawahnya, `text`
// = kalimat ajakan buat caption sosmed.
export type ShareContent = {
  badge: string;
  big: string;
  caption: string;
  detail: string;
  text: string;
};

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:9801/api/v1";

export function isShareKind(kind: string): kind is ShareKind {
  return (SHARE_KINDS as string[]).includes(kind);
}

export function sharePath(nickname: string, kind: ShareKind) {
  return `/s/${encodeURIComponent(nickname)}/${kind}`;
}

// Server-side (page & image route). null = nickname gak ketemu.
export async function fetchShareProfile(nickname: string): Promise<ShareProfile | null> {
  const res = await fetch(`${API_BASE_URL}/public/share/${encodeURIComponent(nickname)}`, {
    next: { revalidate: 60 },
  });
  if (res.status === 404) return null;
  if (!res.ok) throw new Error(`share profile: ${res.status}`);
  return res.json();
}

export function buildShareContent(profile: ShareProfile, kind: ShareKind): ShareContent {
  if (kind === "streak") {
    const days = profile.currentStreak;
    return {
      badge: "🔥 Streak harian",
      big: `${days} hari`,
      caption: days > 0 ? "belajar matematika tanpa bolong" : "siap mulai streak baru",
      detail: `Rekor terpanjang: ${profile.longestStreak} hari`,
      text:
        days > 0
          ? `🔥 Aku udah ${days} hari berturut-turut latihan matematika di MathQuest! Kuat nyusul streak-ku?`
          : `Yuk bareng latihan matematika tiap hari di MathQuest! 🔥`,
    };
  }

  if (kind === "adventure") {
    const adv = profile.adventure;
    if (!adv) {
      return {
        badge: "🗺️ Adventure",
        big: "4.000 soal",
        caption: "dari SD sampai Kampus",
        detail: "40 checkpoint, jatah salah 5x tiap checkpoint",
        text: "Aku lagi mulai Adventure 4.000 soal matematika di MathQuest. Ikutan yuk!",
      };
    }
    const progress = adv.completed ? "Tamat 🏆" : `Checkpoint ${adv.checkpointsCleared}/${adv.totalCheckpoints}`;
    return {
      badge: "🗺️ Ranking Adventure",
      big: `#${adv.rank.rank}`,
      caption: `dari ${adv.rank.total.toLocaleString("id-ID")} petualang`,
      detail: progress,
      text: `🏆 Aku peringkat #${adv.rank.rank} di Adventure MathQuest (${progress})! Berani nyusul?`,
    };
  }

  const tier = profile.tiers.find((t) => t.code === kind);
  const name = tier?.name ?? kind.toUpperCase();
  if (!tier?.rank) {
    return {
      badge: `📚 Jenjang ${name}`,
      big: name,
      caption: "latihan matematika jadi game",
      detail: "Kumpulin poin dari tiap jawaban benar",
      text: `Aku lagi latihan matematika jenjang ${name} di MathQuest. Ikutan yuk!`,
    };
  }
  return {
    badge: `🏆 Ranking ${name}`,
    big: `#${tier.rank.rank}`,
    caption: `dari ${tier.rank.total.toLocaleString("id-ID")} murid ${name}`,
    detail: `${tier.points.toLocaleString("id-ID")} poin`,
    text: `🏆 Aku peringkat #${tier.rank.rank} jenjang ${name} di MathQuest dengan ${tier.points.toLocaleString("id-ID")} poin! Berani nyusul?`,
  };
}

File: services/api/internal/curriculumsvc/share.go
package curriculumsvc

import (
	"context"
	"errors"
	"time"

	"github.com/jackc/pgx/v5"
)

// Share: data publik (tanpa login) buat kartu share + halaman undangan di
// apps/web (/s/{nickname}/{kind}). Dicari lewat nickname, jadi user tanpa
// nickname gak bisa di-share. Rank diambil dari server (bukan dari URL) biar
// gak bisa dipalsuin.
type ShareRank struct {
	Rank  int `json:"rank"`
	Total int `json:"total"`
}

type ShareTier struct {
	Code   string     `json:"code"`
	Name   string     `json:"name"`
	Points int        `json:"points"`
	Rank   *ShareRank `json:"rank"`
}

type ShareAdventure struct {
	CheckpointsCleared int        `json:"checkpointsCleared"`
	TotalCheckpoints   int        `json:"totalCheckpoints"`
	Completed          bool       `json:"completed"`
	TimePercent        float64    `json:"timePercent"`
	Rank               *ShareRank `json:"rank"`
}

type ShareProfile struct {
	Nickname string     `json:"nickname"`
	Avatar   UserAvatar `json:"avatar"`
	// CurrentStreak udah dicek masih nyambung (main hari ini/kemarin), beda
	// sama kolom users.current_streak yang baru di-reset pas main lagi.
	CurrentStreak int             `json:"currentStreak"`
	LongestStreak int             `json:"longestStreak"`
	Tiers         []ShareTier     `json:"tiers"`
	Adventure     *ShareAdventure `json:"adventure"`
}

func (s *Service) GetShareProfile(ctx context.Context, nickname string) (*ShareProfile, error) {
	n := NormalizeNickname(nickname)
	if n == "" {
		return nil, ErrNotFound
	}

	var (
		userID         string
		p              ShareProfile
		lastActiveDate *time.Time
	)
	err := s.db.QueryRow(ctx, `
		SELECT id, username, avatar_type, avatar_key,
			CASE WHEN avatar_type = 'google' THEN google_avatar_url END,
			current_streak, longest_streak, last_active_date
		FROM users WHERE username = $1
	`, n).Scan(&userID, &p.Nickname, &p.Avatar.Type, &p.Avatar.Key, &p.Avatar.GoogleURL,
		&p.CurrentStreak, &p.LongestStreak, &lastActiveDate)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, ErrNotFound
		}
		return nil, err
	}
	yesterday := truncateToUTCDate(time.Now()).AddDate(0, 0, -1)
	if lastActiveDate == nil || truncateToUTCDate(*lastActiveDate).Before(yesterday) {
		p.CurrentStreak = 0
	}

	p.Tiers = []ShareTier{}
	for _, code := range TierLeaderboardCodes {
		board, err := s.GetTierLeaderboard(ctx, userID, code)
		if err != nil {
			return nil, err
		}
		t := ShareTier{Code: code, Name: board.TierName}
		if e := findTierEntry(board); e != nil {
			t.Points = e.Points
			t.Rank = &ShareRank{Rank: e.Rank, Total: board.TotalRanked}
		}
		p.Tiers = append(p.Tiers, t)
	}

	adv, err := s.GetAdventureLeaderboard(ctx, userID)
	if err != nil {
		return nil, err
	}
	if e := findAdventureEntry(adv); e != nil {
		p.Adventure = &ShareAdventure{
			CheckpointsCleared: e.CheckpointsCleared,
			TotalCheckpoints:   adv.TotalCheckpoints,
			Completed:          e.Completed,
			TimePercent:        e.TimePercent,
			Rank:               &ShareRank{Rank: e.Rank, Total: adv.TotalRanked},
		}
	}

	return &p, nil
}

func findTierEntry(b *TierLeaderboard) *TierLeaderboardEntry {
	if b.Me != nil {
		return b.Me
	}
	for i := range b.Entries {
		if b.Entries[i].IsMe {
			return &b.Entries[i]
		}
	}
	return nil
}

func findAdventureEntry(b *AdventureLeaderboard) *AdventureLeaderboardEntry {
	if b.Me != nil {
		return b.Me
	}
	for i := range b.Entries {
		if b.Entries[i].IsMe {
			return &b.Entries[i]
		}
	}
	return nil
}
